client-spark/spark-3/src/main/scala/org/apache/spark/shuffle/celeborn/CelebornShuffleReader.scala

```diff
@@ -106,6 +106,29 @@ class CelebornShuffleReader[K, C](
       }
     }
 
+    // Wait for the shuffle stage end before getting reducer file group.
+    var isShuffleStageEnd: Boolean = false
+    var getStageEndAttempts = 0
+    val getStageEndInterval = conf.clientRpcGetStageEndInterval
+    do {
+      if (getStageEndAttempts > 0) {
+        logInfo(
+          s"Attempting to wait shuffle $shuffleId stage end with interval $getStageEndInterval ms")
+        Thread.sleep(getStageEndInterval)
+      }
+      isShuffleStageEnd =
+        try {
+          shuffleClient.isShuffleStageEnd(shuffleId)
+        } catch {
+          case e: Exception =>
+            logError(s"Failed to check shuffle $shuffleId stage end, assume ended", e)
```

Comment:
It should be fine to assume this,

- the RPC is lightweight and retry with `clientRpcMaxRetries`.